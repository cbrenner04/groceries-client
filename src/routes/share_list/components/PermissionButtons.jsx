import React from 'react';
import PropTypes from 'prop-types';

const PermissionButtons = props => (
  <div>
    <h3 className="text-capitalize">{props.status}</h3>
    <br />
    {
      props.users.map(({ user, users_list: { id, permissions } }) => {
        if (user.id === props.userId) return '';
        if (props.userIsOwner) {
          return (
            <button
              key={id}
              id={`${props.status}-user-${user.id}`}
              className={'list-group-item list-group-item-action d-flex justify-content-between align-items-center'
                + 'btn btn-link'}
              onClick={() => props.togglePermission(id, permissions, props.status)}
            >
              <span>{user.email}</span>
              <span
                id={`perm-${permissions}`}
                className={`badge badge-${permissions === 'write' ? 'success' : 'primary'}`}
              >
                {permissions}
              </span>
            </button>
          );
        }
        return <div key={id} id={`${props.status}-user-${user.id}`} className="list-group-item">{user.email}</div>;
      })
    }
    <br />
  </div>
);

PermissionButtons.propTypes = {
  togglePermission: PropTypes.func.isRequired,
  userIsOwner: PropTypes.bool.isRequired,
  userId: PropTypes.number.isRequired,
  status: PropTypes.string.isRequired,
  users: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    email: PropTypes.string,
  }).isRequired).isRequired,
};

export default PermissionButtons;
