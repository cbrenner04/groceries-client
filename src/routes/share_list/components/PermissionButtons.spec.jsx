import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import PermissionButtons from './PermissionButtons';

describe('PermissionButtons', () => {
  let props;

  beforeEach(() => {
    props = {
      togglePermission: jest.fn(),
      userIsOwner: true,
      userId: 1,
      status: 'accepted',
      users: [
        {
          user: {
            id: 1,
            email: 'foo@example.com',
          },
          users_list: {
            id: 1,
            permissions: 'write',
          },
        },
        {
          user: {
            id: 2,
            email: 'bar@example.com',
          },
          users_list: {
            id: 2,
            permissions: 'write',
          },
        },
        {
          user: {
            id: 3,
            email: 'baz@example.com',
          },
          users_list: {
            id: 4,
            permissions: 'read',
          },
        },
      ],
    };
  });

  it('renders read and write badges when user is owner', () => {
    props.userIsOwner = true;
    const { container, getByTestId, queryByTestId } = render(<PermissionButtons {...props} />);

    expect(container).toMatchSnapshot();
    expect(queryByTestId('accepted-user-1')).toBeNull();
    expect(getByTestId('accepted-user-2').firstChild.children[1]).toHaveAttribute('data-test-id', 'perm-write');
    expect(getByTestId('accepted-user-2').firstChild.children[1]).toHaveClass('badge-success');
    expect(getByTestId('accepted-user-3').firstChild.children[1]).toHaveAttribute('data-test-id', 'perm-read');
    expect(getByTestId('accepted-user-3').firstChild.children[1]).toHaveClass('badge-primary');
  });

  it('does not render read and write badges when user is not owner', () => {
    props.userIsOwner = false;
    const { container, getByTestId, queryByText } = render(<PermissionButtons {...props} />);

    expect(container).toMatchSnapshot();
    expect(queryByText('foo@example.com')).toBeNull();
    expect(getByTestId('accepted-user-2')).toHaveTextContent('bar@example.com');
    expect(getByTestId('accepted-user-2')).not.toHaveTextContent('write');
    expect(getByTestId('accepted-user-3')).toHaveTextContent('baz@example.com');
    expect(getByTestId('accepted-user-3')).not.toHaveTextContent('read');
  });

  it('toggles user permissions on click', async () => {
    props.userIsOwner = true;
    const { getByTestId } = render(<PermissionButtons {...props} />);

    fireEvent.click(getByTestId('accepted-user-2').firstChild);
    expect(props.togglePermission).toHaveBeenCalledWith(2, 'write', 'accepted');
  });
});
