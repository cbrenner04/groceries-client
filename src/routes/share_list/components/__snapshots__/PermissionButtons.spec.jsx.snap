// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionButtons does not render read and write badges when user is not owner 1`] = `
<div>
  <h3
    class="text-capitalize"
  >
    accepted
  </h3>
  <br />
  <div
    class="list-group"
  >
    
    <div
      data-test-id="accepted-user-2"
    >
      <div
        class="list-group-item"
      >
        bar@example.com
      </div>
    </div>
    <div
      data-test-id="accepted-user-3"
    >
      <div
        class="list-group-item"
      >
        baz@example.com
      </div>
    </div>
  </div>
  <br />
</div>
`;

exports[`PermissionButtons renders read and write badges when user is owner 1`] = `
<div>
  <h3
    class="text-capitalize"
  >
    accepted
  </h3>
  <br />
  <div
    class="list-group"
  >
    
    <div
      data-test-id="accepted-user-2"
    >
      <button
        class="d-flex justify-content-between align-items-center list-group-item list-group-item-action"
      >
        <span>
          bar@example.com
        </span>
        <span
          class="badge badge-success"
          data-test-id="perm-write"
        >
          write
        </span>
      </button>
    </div>
    <div
      data-test-id="accepted-user-3"
    >
      <button
        class="d-flex justify-content-between align-items-center list-group-item list-group-item-action"
      >
        <span>
          baz@example.com
        </span>
        <span
          class="badge badge-primary"
          data-test-id="perm-read"
        >
          read
        </span>
      </button>
    </div>
  </div>
  <br />
</div>
`;
